### PR TITLE
feat(replay): Wrap rrweb library in react context

### LINKED ISSUE #33519: Separate basic controls (forward, back, 2x, etc.) into separate components
See https://github.com/rrweb-io/rrweb/blob/master/docs/recipes/customize-replayer.md

### PR DESCRIPTION
**Pull request by ryan953**
This change implements a react-centric replacement for the `<BaseRRWebReplayer>` component, and the `rrweb-player` npm package. That package is based on svelte and doesn't really do things in a react way. 

So what we've got instead are the core features of a react-context based system.
1. There's the ReplayPlayer, which wraps `rrweb` and hosts the iframe where replays will be injected. 
2. We've also got the ReplayController, which is where the play/pause buttons are hosted, as well as the video scrobber and so on. 

Using context to manage the replay player is important because it allows us to move child components around inside the render tree and still give them access to the replay data.

Fixes: #33519 #33513

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -129,6 +129,7 @@
     "react-virtualized": "^9.22.3",
     "reflux": "0.4.1",
     "regenerator-runtime": "^0.13.3",
+    "rrweb": "^1.1.3",
     "rrweb-player": "^0.7.13",
```
Comment (ryan953) [inline package.json:133]: The `rrweb-player` package depends on `rrweb`: https://github.com/rrweb-io/rrweb/blob/master/packages/rrweb-player/package.json#L28 
The `*-player` package uses svelte to add play/pause buttons and the like around the base `rrweb` iframe.

I'm bringing in `rrweb` explicitly so we can wrap the iframe ourselves, and use buttons and icons from the sentry component library instead. Also we'll have fewer svelte and a better react-oriented way to interact with the rrweb class.

File: static/app/components/replays/replayController.tsx [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+              title={isPlaying ? t('Pause the Replay') : t('Play the Replay')}
+              icon={isPlaying ? <IconPause size="sm" /> : <IconPlay size="sm" />}
+              onClick={() => togglePlayPause(!isPlaying)}
+              aria-label={t('Toggle real-time updates')}
+            />
+            <Button
+              data-test-id="replay-forward-10s"
+              size="xsmall"
+              title={t('Go forwards 10 seconds')}
+              icon={<IconRefresh color="gray500" size="sm" />}
```
Comment (ryan953) [inline static/app/components/replays/replayController.tsx:91]: This is the wrong icon, but i couldn't find a clock-wise arrow in the storybook examples

This is what it looks like now:
Screen Shot 2022-04-08 at 12 02 37 PM

File: static/app/components/replays/replayController.tsx [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+  | 'unset';
+
+const FlexLayout = styled('div')<{
+  alignItems?: AlignItemsValues;
+  direction?: FlexDirectionValues;
+  gap?: Parameters<typeof space>[0];
+}>`
+  display: flex;
+  flex-direction: ${p => p.direction ?? 'initial'};
+  gap: ${p => space(p.gap ?? 1)};
```
Comment (billyvg) [inline static/app/components/replays/replayController.tsx:175]: We've been avoiding using `gap` for flexbox (opting only for grid) as we still have a small amt of users that use an unsupported browser https://caniuse.com/flexbox-gap

File: static/app/components/replays/replayPlayer.tsx [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+  // Create the `rrweb` instance which creates an iframe inside `viewEl`
+  useEffect(() => initRoot(viewEl.current), [viewEl.current]);
+
+  // Read the initial width/height where the player will be inserted, to setup view transform/scaling
+  // Also listen for resize events and so we can update view transform/scaling
+  useEffect(() => {
+    const handleResize = () =>
+      setWindowDimensions(windowEl.current?.getBoundingClientRect());
+    handleResize();
+    window.addEventListener('resize', handleResize);
```
Comment (billyvg) [inline static/app/components/replays/replayPlayer.tsx:34]: Should this be throttled/debounced? or should we use resize observer?

File: static/app/components/replays/replayPlayer.tsx [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+  .replayer-mouse-tail {
+    position: absolute;
+    pointer-events: none;
+  }
+
+  /* Skin the player to match our aesthetic */
+  iframe {
+    border: none;
+  }
+  .replayer-mouse {
```
Comment (billyvg) [inline static/app/components/replays/replayPlayer.tsx:71]: I liked the glowy circle that we currently have, but nb

**Review by billyvg — COMMENTED** [review]
👏

File: static/app/components/replays/replayController.tsx [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+        showLabel={false}
+      />
+      <ButtonLayout>
+        <FlexLayout direction="row" alignItems="center">
+          <ButtonBar merged>
+            <Button
+              data-test-id="replay-back-10s"
+              size="xsmall"
+              title={t('Go back 10 seconds')}
+              icon={<IconRefresh color="gray500" size="sm" />}
```
[the remark this answers is not in the document]
Comment (ryan953) [reply static/app/components/replays/replayController.tsx:74]: yup

Screen Shot 2022-04-08 at 2 38 57 PM

I hadn't checked it earlier, only copy+pasted from the storybook icon page.

File: static/app/components/replays/replayController.tsx (hunk 3 above)
Comment (ryan953) [reply static/app/components/replays/replayController.tsx:175]: ahh, i can do the css another way. This'll also impact the comment above regarding `FlexLayout` so i'll see how the whole area shakes out

File: static/app/components/replays/replayPlayer.tsx (hunk 4 above)
Comment (ryan953) [reply static/app/components/replays/replayPlayer.tsx:34]: Resize observer could be cool, i didn't know about that before!

I didn't measure the impact of adding throttle or debounce, I can follow up on this diff with that I think.

File: static/app/components/replays/replayPlayer.tsx (hunk 5 above)
Comment (ryan953) [reply static/app/components/replays/replayPlayer.tsx:71]: Fair enough, I ripped it all out to highlight just the basics that were needed to get it working. 

I can bring the colors back and keep it separate from this core player, to make it easier for design to parse out what we've added.

File: static/app/components/replays/replayController.tsx [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+import useRAF from './useRAF';
+import {formatTime} from './utils';
+
+const SECOND = 1000;
+
+function useCurrentTime(callback: () => number) {
+  const [currentTime, setCurrentTime] = useState(0);
+  useRAF(() => setCurrentTime(callback));
+  return currentTime;
+}
```
[the remark this answers is not in the document]
Comment (ryan953) [reply static/app/components/replays/replayController.tsx:22]: Yeah, it'll need to go in there really soon.

I'll check that consumers are not re-rendering like crazy and move it up

[the remark this answers is not in the document]
Comment (ryan953) [reply static/app/components/replays/replayController.tsx]: moved it into context

File: static/app/components/replays/replayController.tsx [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+      <ButtonLayout>
+        <FlexLayout direction="row" alignItems="center">
+          <ButtonBar merged>
+            <Button
+              data-test-id="replay-back-10s"
+              size="xsmall"
+              title={t('Go back 10 seconds')}
+              icon={<IconRefresh color="gray500" size="sm" />}
+              onClick={() => setCurrentTime(currentTime - 10 * SECOND)}
+              aria-label={t('Jump back 10 seconds')}
```
[the remark this answers is not in the document]
Comment (ryan953) [reply static/app/components/replays/replayController.tsx]: no reason, fixed all the labels.

File: static/app/components/replays/replayController.tsx [hunk 9]
```diff
@@ -0,0 +1,236 @@
...
+  | 'unsafe center'
+  | 'inherit'
+  | 'initial'
+  | 'revert'
+  | 'revert-layer'
+  | 'unset';
+
+const FlexLayout = styled('div')<{
+  alignItems?: AlignItemsValues;
+  direction?: FlexDirectionValues;
```
[the remark this answers is not in the document]
Comment (ryan953) [reply static/app/components/replays/replayController.tsx]: Got rid of all this and used a grid instead. Much less code, but it'll be slightly annoying if/when someone adds a new button to the row -> they will need to add a size into the `grid-template-columns` list.

File: static/app/components/replays/replayPlayer.tsx (hunk 5 above)
Comment (ryan953) [reply static/app/components/replays/replayPlayer.tsx:85]: brought back all the mouse css

File: static/app/components/replays/replayPlayer.tsx (hunk 4 above)
Comment (ryan953) [reply static/app/components/replays/replayPlayer.tsx]: don't need to follow up.

I found `useResizeObserver` which is cool, and also replaced `getBoundingClientRect` with reads to `.clientWidth` and `.clientHeight`, so the browser doesn't need to calculate all the values.

File: static/app/components/replays/replayContext.tsx [hunk 10]
```diff
@@ -0,0 +1,243 @@
...
+  // if (replayerRef.current) {
+  //   console.log({
+  //     replayerRef,
+  //     currentTime: replayerRef.current.getCurrentTime(),
+  //     metadata: replayerRef.current.getMetaData(),
+  //     mirror: replayerRef.current.getMirror(),
+  //     timeOffset: replayerRef.current.getTimeOffset(),
+  //     config: replayerRef.current.config,
+  //   });
+  // }
```
Comment (billyvg) [inline static/app/components/replays/replayContext.tsx:217]: Can we remove this?

File: static/app/components/replays/replayController.tsx [hunk 11]
```diff
@@ -0,0 +1,194 @@
...
+            onClick={() => togglePlayPause(!isPlaying)}
+            aria-label={isPlaying ? t('Pause the Replay') : t('Play the Replay')}
+          />
+          <Button
+            data-test-id="replay-forward-10s"
+            size="xsmall"
+            title={t('Go forwards 10 seconds')}
+            icon={<IconRefresh color="gray500" size="sm" />}
+            onClick={() => setCurrentTime(currentTime + 10 * SECOND)}
+            aria-label={t('Go forwards 10 seconds')}
```
Comment (billyvg) [inline static/app/components/replays/replayController.tsx:84]: I think

```suggestion
            title={t('Go forward 10 seconds')}
            icon={<IconRefresh color="gray500" size="sm" />}
            onClick={() => setCurrentTime(currentTime + 10 * SECOND)}
            aria-label={t('Go forward 10 seconds')}
```

File: static/app/components/replays/replayController.tsx [hunk 12]
```diff
@@ -0,0 +1,194 @@
...
+          ))}
+        </ButtonBar>
+
+        <Button
+          data-test-id="replay-fullscreen"
+          size="xsmall"
+          title={t('View the Replay in full screen')}
+          icon={<IconResize size="sm" />}
+          onClick={onFullscreen}
+          aria-label={t('View in full screen')}
```
Comment (billyvg) [inline static/app/components/replays/replayController.tsx:121]:
```suggestion
          title={t('View in full screen')}
          icon={<IconResize size="sm" />}
          onClick={onFullscreen}
          aria-label={t('View in full screen')}
```

**Review by billyvg — APPROVED** [review]

File: static/app/components/replays/replayContext.tsx (hunk 10 above)
Comment (ryan953) [reply static/app/components/replays/replayContext.tsx]: yes, gone.

**Pull request merged**